### PR TITLE
Implement bit conversion for QMontgomeryUInt, fix dtype, symbolic decomposition and classical action of modular gates, and add classical action tests

### PR DESCRIPTION
**Pull request by NoureldinYosri**
- Implement bit conversion for QMontgomeryUInt: numbers in montgomery form can be treated as just UInts when it comes to bits.
- dtype of modular bloqs must be QMontgomeryUInt not QUInt

This reproduces most of the numbers in the table in Fig 8 of https://arxiv.org/abs/2306.08585. The remaining bloqs are the different multiplication bloqs and the controlled version of bloqs

### CODE REVIEW

File: qualtran/_infra/data_types.py [hunk 1]
```diff
@@ -787,10 +787,11 @@ def get_classical_domain(self) -> Iterable[Any]:
         return range(2**self.bitsize)
 
     def to_bits(self, x: int) -> List[int]:
-        raise NotImplementedError(f"to_bits not implemented for {self}")
+        self.assert_valid_classical_val(x)
+        return [int(x) for x in f'{int(x):0{self.bitsize}b}']
 
     def from_bits(self, bits: Sequence[int]) -> int:
-        raise NotImplementedError(f"from_bits not implemented for {self}")
+        return int("".join(str(x) for x in bits), 2)
```
Comment (mpharrigan) [inline qualtran/_infra/data_types.py:794]: add a simple test?

File: qualtran/bloqs/arithmetic/comparison.py [hunk 2]
```diff
@@ -875,6 +875,25 @@ def wire_symbol(
             return TextBox('t⨁(a>b)')
         raise ValueError(f'Unknown register name {reg.name}')
 
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        if self.bitsize == 1:
+            return {(MultiControlX(cvs=(1, 0)), 1)}
+        # Ignore Allocation and Deallocation bloqs for now.
```
Comment (mpharrigan) [inline qualtran/bloqs/arithmetic/comparison.py:881]: I think we typically ignore these in `build_call_graph`, so not worth noting.

Could you add a quick `qlt_testing.assert_equivalent_bloq_counts` test?

File: qualtran/bloqs/factoring/mod_sub.py [hunk 3]
```diff
@@ -94,6 +97,15 @@ def build_composite_bloq(self, bb: 'BloqBuilder', x: Soquet, y: Soquet) -> Dict[
     def pretty_name(self) -> str:
         return f'y = y - x mod {self.p}'
 
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        # Ignore allocation, deallocation, splits and joins for now.
```
Comment (mpharrigan) [inline qualtran/bloqs/factoring/mod_sub.py:101]: don't necessarily need to note this.

can you add a qlt_testing.assert_equilvalent_bloq_counts?

**Review by mpharrigan — APPROVED** [review]
a couple requests for tests, but looking good

File: qualtran/_infra/data_types.py (hunk 1 above)
Comment (NoureldinYosri) [reply qualtran/_infra/data_types.py:794]: done

File: qualtran/bloqs/arithmetic/comparison.py (hunk 2 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: done

File: qualtran/bloqs/factoring/mod_sub.py (hunk 3 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/factoring/mod_sub.py]: done

**Pull request merged**